File: app/src/main/java/com/estebanlamas/planesradar/presentation/map/MapViewModel.java
package com.estebanlamas.planesradar.presentation.map;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.ViewModel;

import com.estebanlamas.planesradar.data.AircraftRepository;
import com.estebanlamas.planesradar.presentation.model.Aircraft;

import java.util.List;
import java.util.Timer;
import java.util.TimerTask;

/**
 * Created by esteban on 24/9/17
 */

public class MapViewModel extends ViewModel {
    private static final int REFRESH_MILLISECONDS = 3000;
    private LiveData<List<Aircraft>> aircraftList;
    private Timer refreshTimer = new Timer();

    public MapViewModel(AircraftRepository repository) {
        if(aircraftList!=null) {
            return;
        }
        initRadar(repository);
    }

    LiveData<List<Aircraft>> getAircraftList() {
        return aircraftList;
    }

    private void initRadar(final AircraftRepository repository) {
        refreshTimer.schedule(new TimerTask() {
            @Override
            public void run() {
                requestAircrafts(repository);
            }
        },0,REFRESH_MILLISECONDS);
    }

    private void requestAircrafts(AircraftRepository repository) {
        aircraftList = repository.getAircraftList();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        refreshTimer.cancel();
    }
}

File: app/src/main/java/com/estebanlamas/planesradar/presentation/map/MapViewModel.kt
package com.estebanlamas.planesradar.presentation.map

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.asLiveData
import com.estebanlamas.planesradar.domain.GetAircraftsUseCase
import com.estebanlamas.planesradar.domain.model.Aircraft
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.map
import javax.inject.Inject

/**
 * Created by esteban on 24/9/17
 */

class MapViewModel
@Inject constructor(getAircraftsUseCase: GetAircraftsUseCase): ViewModel() {
    private val _error = MutableLiveData<String>()
    val error: LiveData<String> = _error
    val aircraftList: LiveData<List<Aircraft>> = getAircraftsUseCase.execute()
        .catch { _error.postValue(it.message) }
        .map { it.aircrafts }
        .asLiveData()
}
